### PR TITLE
feat(h706): Advanced carry forward

### LINKED ISSUE #706: Advanced Carry Forward
### Goal

Add an **Advanced Carry Forward** configuration section to all multiple-choice question types in Survey Creator (Checkbox, Radiogroup, Dropdown, etc.) so survey builders can aggregate, deduplicate, and manage choices carried forward from multiple source questions directly on the target question—without relying on a separate helper question.

### Scope

**In scope**

- A dedicated **Advanced Carry Forward** panel in the Survey Creator property grid for all multiple-choice question types.
- Multi-select of source questions (multiple-choice / multiple-select types) whose choices feed the target question.
- Runtime aggregation of choices from selected sources, with deduplication by value and text so each choice appears once while preserving original labels and underlying values.
- **Priority choices** that always stay at the top of the list and are never randomized.
- Non-priority carried-forward choices respect the target question’s existing randomization settings.
- **Max-selection** support: inherit limits from source-question logic and/or allow the creator to set a dynamic limit for the aggregated list (`maxSelectedChoices`).

**Out of scope**

- Carry forward from non-multiple-choice question types (unless explicitly added in a follow-up).
- Server-side/API changes (form definition is stored as SurveyJS JSON; behavior is client-side in Creator and runtime).

### Implementation Notes

- Follow existing Hub SurveyJS extension patterns under `hub/lib/survey-features/*` (Creator property registration, runtime hooks, pure aggregation logic, Vitest coverage).
- Prefer a linear pipeline for choice reconciliation: **extract → deduplicate → apply priority ordering → apply randomization rules → apply max-selection → merge into target choices**.
- Use `Map`/`Set` for O(1) deduplication and cross-referencing source choices by value/text.
- Isolate side effects: build the final choice list in memory, then apply to the target question in one step.
- Reuse or align with existing choice-extraction patterns where applicable (e.g. `getLoopChoicesFromQuestion` in question-loops) without coupling Advanced Carry Forward to loop machinery.
- Ensure behavior is consistent in **survey runtime** and **Survey Creator preview**.

### Acceptance Criteria

- [ ] **Creator UI.** Multiple-choice questions expose an **Advanced Carry Forward** section where the builder can select one or more eligible source questions.
- [ ] **Aggregation.** At runtime, choices from all selected sources are merged into the target question’s choice list.
- [ ] **Deduplication.** Identical choices (same value and text) appear only once; labels and values match the source definitions.
- [ ] **Priority choices.** Configured priority choices remain fixed at the top and are excluded from randomization.
- [ ] **Randomization.** Non-priority carried-forward choices follow the target question’s randomization setting.
- [ ] **Max selection.** The target question enforces `maxSelectedChoices` per configuration—either inherited from source logic or set explicitly for the aggregated list.
- [ ] **Preview parity.** Creator preview matches survey-taking behavior for aggregation, ordering, randomization, and max-selection.
- [ ] **Test coverage.** Unit tests cover aggregation, deduplication, priority ordering, randomization interaction, and max-selection edge cases (empty sources, overlapping choices, single vs multiple sources).

### PR DESCRIPTION
**Pull request by orilux**
## Summary

- Adds **Advanced carry forward** as a Hub SurveyJS extension (`ENDATIX_ENABLE_EXTENSIONS=true`) for SelectBase types: checkbox, radiogroup, dropdown, tagbox, imagepicker, ranking, buttongroup.
- Lets builders copy choices from **multiple source questions** with **All / Selected / Unselected** modes, deduplication, optional **priority items**, and a **maximum number of choices** cap (list size, not selections).
- Runtime syncs destination `choices` when source answers or carry-forward settings change; supports chained carry-forward targets via dependency ordering.
- Mutual exclusion with data lists, choices-by-URL, and native **Copy choices from** — enforced in Creator UI and at runtime via `isAdvancedCarryForwardEnabled`.
- Reuses question-loops utilities (`extractUniqueChoicesBy`, `getLoopChoicesFromQuestion`) and group-randomization for priority pinning when **Choices order → Random**.
- Property hints (`pehelp`) on all carry-forward fields, including the main toggle.

## Deviations from original UI / product plan

**Separate tab → Choices tab**

- Original plan: dedicated Serializer category `advancedCarryForward`, custom tab icon/title, property-grid decoration in `creator-bindings`.
- Shipped: all settings live on the existing **Choices** tab (`category: 'choices'`), alongside where builders already configure choice sources.
- When advanced carry forward is on, native **Copy choices from** and the inline **choices** editor are hidden (same UX pattern as native copy-from).
- Mode uses SurveyJS native segmented control (`all` / `selected` / `unselected`) with label **“Which choice options to copy”**, not loop-style dropdown labels (`Selected Only`, etc.). Legacy JSON labels still map at runtime.

**Removed max-selection inheritance (not max list size)**

- Original plan: `advancedCarryForwardMaxMode` (inherit / manual / unlimited) and `advancedCarryForwardMaxSelected`, with runtime overwrite of `maxSelectedChoices`.
- Shipped: **removed entirely**. Builders use SurveyJS **Max selected choices** on the General tab as today; carry-forward does not touch it.
- Rationale: weak business value for multi-source forms and duplicated SurveyJS behavior; inherit logic was confusing and niche.
- Kept instead: **`advancedCarryForwardMaxChoices`** — caps how many options **appear** in the carried-forward list (aligned with question loops’ `maxLoopCount`; `0` = unlimited; priority items always included).

**Other plan deltas**

- End-user docs (original PR3) deferred.
- Custom PropertyGrid editor for multi-source picker deferred (Serializer `multiplevalues` first).
- Ranking as a source for Selected/Unselected: documented in mode help as behaving differently from checkbox-style selection.

## Implementation notes

- Wired via `ExtensionModule` in `core-registry`; shared Creator binding helper with blind-search-tagbox.
- Sync hooks: `onValueChanged` + question `onPropertyChanged` (Creator property edits do not fire value-changed).
- Review hardening: type-safe mode parsing, re-entrancy `try/finally`, image metadata preservation, raw-value enrichment after dedup, registry `visibleIf` wrap-once + restore for tests.

## Related

Closes #706 

## Test plan

- [ ] Enable `ENDATIX_ENABLE_EXTENSIONS=true`, restart dev server; open form designer on a checkbox → **Choices** tab shows **Advanced carry forward** toggle and hint.
- [ ] Toggle on → set multiple sources, switch All/Selected/Unselected; preview updates when source answers change.
- [ ] Set **Priority items** + **Choices order → Random** → priority stays fixed, rest shuffles.
- [ ] Set **Maximum number of choices** → list capped; priority items still shown.
- [ ] Confirm mutual exclusion: data list / choices-by-URL / native copy-from hide carry-forward (and vice versa).
- [ ] Tagbox with blind search + carry forward: choices sync; blind search filtering still works.
- [ ] Import JSON with conflicting flags (e.g. data list + carry forward) → carry-forward sync does not run.
- [ ] `pnpm vitest run lib/survey-features/advanced-carry-forward` passes.
## Related Issues
List any related issues (e.g., fixes #123, closes #456).

## Type of Change
Check all that apply.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Dependency update
- [ ] Refactoring (non-breaking change which improves code quality)
- [ ] Other (please describe)

## Checklist
- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules

> [!NOTE]
> - [x] I confirm that my Pull Request follows all the checklist requirements above

## Screenshots
image

## Summary by CodeRabbit

* **New Features**
  * Added Advanced Carry Forward support for survey questions, including source selection, priority choices, and a choice limit.
  * Enabled automatic syncing so carried-forward choices update when source answers or related settings change.
  * Added creator-facing help text and visibility rules so the new settings appear only when appropriate.

* **Bug Fixes**
  * Improved handling of choice lists, including preserving item details, keeping valid selections, and avoiding duplicate or stale updates.
  * Updated Data List visibility so it no longer conflicts with the new carry-forward setup.

### CODE REVIEW

File: lib/survey-features/infrastructure/creator-survey-bindings.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  const handleSurveyInstanceCreated = (
+    _: unknown,
+    options: SurveyInstanceCreatedEvent,
+  ) => {
+    if (options.area === PROPERTY_GRID_AREA) {
+      return;
+    }
+
+    bindSurveyForCreatorArea(options.area, options.survey);
+  };
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/infrastructure/creator-survey-bindings.ts:57]: Addressed 🥕

File: lib/survey-features/advanced-carry-forward/use-cases/sync-carry-forward-target.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+      target.value = prunedValue;
+    }
+
+    return;
+  }
+
+  if (!isChoiceValueValid(validValues, currentValue)) {
+    target.clearValue();
+  }
+}
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/advanced-carry-forward/use-cases/sync-carry-forward-target.ts]: Addressed 🥕

File: lib/survey-features/advanced-carry-forward/use-cases/load-carry-forward-targets.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...

+  ADVANCED_CARRY_FORWARD_PRIORITY_ITEMS_PROPERTY,

+  ADVANCED_CARRY_FORWARD_MAX_CHOICES_PROPERTY,

+]);

+

+let isUpdatingCarryForward = false;
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/advanced-carry-forward/use-cases/load-carry-forward-targets.ts]: Addressed 🥕

File: lib/survey-features/advanced-carry-forward/use-cases/carry-forward-question-utils.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
+import type { Question } from 'survey-core';
+import { isSelectBaseQuestion } from '@/lib/survey-features/question-loops/loop-utils';
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/advanced-carry-forward/use-cases/carry-forward-question-utils.ts]: Addressed 🥕

File: lib/survey-features/advanced-carry-forward/types.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import type { QuestionSelectBase } from 'survey-core';
+import type { AdvancedCarryForwardModeValue } from './carry-forward-mode-values';
+
+export interface AdvancedCarryForwardQuestion extends QuestionSelectBase {
+  advancedCarryForwardEnabled?: boolean;
+  advancedCarryForwardSources?: string[];
+  advancedCarryForwardMode?: AdvancedCarryForwardModeValue | string;
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/advanced-carry-forward/types.ts]: Addressed 🥕

File: lib/survey-features/advanced-carry-forward/infrastructure/registry.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...

+        propertyName,

+      ),

+    ),

+  );

+}
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/advanced-carry-forward/infrastructure/registry.ts]: It may be trivial but your word is my command. Addressed 🥕

File: lib/survey-features/advanced-carry-forward/infrastructure/registry.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...

+

+const NATIVE_CHOICE_PROPERTIES_TO_HIDE = [

+  'choicesFromQuestion',

+  'choices',

+] as const;
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/advanced-carry-forward/infrastructure/registry.ts]: Addressed 🥕

File: lib/survey-features/advanced-carry-forward/infrastructure/creator-help.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+export function resetAdvancedCarryForwardCreatorHelpForTests(): void {
+  const translations = getLocaleStrings('en');
+
+  delete translations.pehelp[ADVANCED_CARRY_FORWARD_ENABLED_PROPERTY];
+  delete translations.pehelp[ADVANCED_CARRY_FORWARD_MODE_PROPERTY];
+  delete translations.pehelp[ADVANCED_CARRY_FORWARD_PRIORITY_ITEMS_PROPERTY];
+  delete translations.pehelp[ADVANCED_CARRY_FORWARD_MAX_CHOICES_PROPERTY];
+
+  isCreatorHelpRegistered = false;
+}
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/advanced-carry-forward/infrastructure/creator-help.ts]: Addressed 🥕

File: lib/survey-features/carry-forward/carry-forward-properties.ts [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+  name: ADVANCED_CARRY_FORWARD_MAX_CHOICES_PROPERTY,
+  dependsOn: [ADVANCED_CARRY_FORWARD_ENABLED_PROPERTY],
+  displayName: 'Maximum number of choices',
+  category: ADVANCED_CARRY_FORWARD_CHOICES_CATEGORY,
+  type: 'number',
+  default: 0,
+  minValue: 0,
+  visibleIndex: 5,
+  visibleIf: isCarryForwardFeatureVisible,
+};
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/carry-forward/carry-forward-properties.ts:160]: Addressed 🥕

File: lib/survey-features/advanced-carry-forward/__tests__/survey-bindings.test.ts [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    const model = new Model(carryForwardSurveyJson);
+    const firstDispose = bindAdvancedCarryForwardToSurvey(model);
+
+    // Act
+    const secondDispose = bindAdvancedCarryForwardToSurvey(model);
+
+    // Assert
+    expect(secondDispose).toBeTypeOf('function');
+    firstDispose();
+  });
```
[the remark this answers is not in the document]
Comment (orilux) [reply lib/survey-features/advanced-carry-forward/__tests__/survey-bindings.test.ts]: Addressed 🥕

File: lib/survey-features/advanced-carry-forward/use-cases/load-carry-forward-targets.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+export function loadCarryForwardTargets(
+  sender: SurveyModel,
+  options: ValueChangedEvent,
+): void {
+  if (updatingCarryForwardSurveys.has(sender)) {
+    return;
+  }
+
+  const carryForwardTargets = getAllCarryForwardTargets(sender);
+  const targetsToUpdate = carryForwardTargets.filter(
```
Comment (PeterMarinov) [inline lib/survey-features/advanced-carry-forward/use-cases/load-carry-forward-targets.ts]: When source X changes, only targets whose advancedCarryForwardSources directly includes X are synced. If A sources from X and B sources from A, changing X syncs A but not B. Worse: pruneInvalidSelections inside syncSingleCarryForwardTarget may set target.value during sync, firing onValueChanged while the re-entrancy guard (updatingCarryForwardSurveys) is active — those nested events are dropped entirely.

Fix: Collect all targets transitively downstream of the changed question (walk dependency graph), or sync all carry forward targets in dependency order on any relevant value change (Helpers.isArraysEqual makes no-op cheap). Add a regression test: X → A → B, change X, assert B.choices updates.

File: lib/survey-features/advanced-carry-forward/use-cases/sync-carry-forward-target.ts [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+} from './limit-carry-forward-choices';
+
+const PRIORITY_GROUP = 'priority';
+
+function buildSourceChoiceIndex(
+  sourceQuestions: ReturnType<typeof getCarryForwardSourceQuestions>,
+): Map<unknown, ItemValue> {
+  const index = new Map<unknown, ItemValue>();
+
+  for (const source of sourceQuestions) {
```
Comment (PeterMarinov) [inline lib/survey-features/advanced-carry-forward/use-cases/sync-carry-forward-target.ts]: SurveyJS native carry forward reads source.visibleChoices, skips built-in items via isBuiltInChoice, and uses isItemSelected for selected/unselected modes ([getChoicesFromSelectQuestion](https://github.com/surveyjs/survey-library/blob/master/packages/survey-core/src/question_baseselect.ts#L1659)). This implementation reads raw question.choices, which is wrong when the source uses native CF, data lists, choicesVisibleIf, or is itself a carry-forward target whose runtime choices live in visibleChoices.

Fix: New use-case getChoicesFromSourceQuestion(source, mode) mirroring SurveyJS semantics; stop routing carry-forward through getLoopChoicesFromQuestion.

File: lib/survey-features/advanced-carry-forward/infrastructure/survey-bindings.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+    const handler = (_: unknown, options: { name: string }) => {
+      loadCarryForwardTargetsFromPropertyChange(model, options.name);
+    };
+
+    question.onPropertyChanged.add(handler);
+    questionHandlers.set(question, handler);
+  };
+
+  model.getAllQuestions().forEach(attachToQuestion);
```
Comment (PeterMarinov) [inline lib/survey-features/advanced-carry-forward/infrastructure/survey-bindings.ts]: Attaches a property-change handler to all questions, not just carry-forward targets. In Creator this scales O(questions × property edits) and fires syncAllCarryForwardTargets for unrelated edits that happen to touch registered props on any question.

Fix: Attach handlers only to Carry Forward -enabled targets (and re-attach on `onQuestionAdded` when target becomes enabled). Better: drop broad property listeners once addDependedQuestion + scoped config sync is wired.

File: lib/survey-features/carry-forward/infrastructure/carry-forward-sync.ts [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+export function loadCarryForwardTargetsFromPropertyChange(
+  survey: SurveyModel,
+  propertyName: string,
+): void {
+  if (!CARRY_FORWARD_CONTROL_PROPS.has(propertyName)) {
+    return;
+  }
+
```
Comment (PeterMarinov) [inline lib/survey-features/carry-forward/infrastructure/carry-forward-sync.ts:83]: Any carry forward control property change on any question calls `syncAllCarryForwardTargets`. Should scope to the changed target + downstream dependents only.

Fix: Pass changed question into sync helper; use orderCarryForwardTargetsByDependencies on { changedTarget } downstream.

File: lib/survey-features/advanced-carry-forward/infrastructure/survey-bindings.ts [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    return () => {};
+  }
+
+  modelWithFlags[ADVANCED_CARRY_FORWARD_HANDLERS_ATTACHED_KEY] = true;
+
+  getCarryForwardTargetsInDependencyOrder(model).forEach((target) => {
+    syncSingleCarryForwardTarget(model, target);
+  });
+
+  model.onValueChanged.add(loadCarryForwardTargets);
```
Comment (PeterMarinov) [inline lib/survey-features/advanced-carry-forward/infrastructure/survey-bindings.ts]: Reimplements propagation with model.onValueChanged + manual filtering. SurveyJS already tracks dependents via `addDependedQuestion` / `updateDependedQuestions` / `updateDependedQuestion` on [QuestionSelectBase](https://github.com/surveyjs/survey-library/blob/master/packages/survey-core/src/question_baseselect.ts).

Fix: On bind, for each target→source edge call source.addDependedQuestion(target); wrap target sync to run when updateDependedQuestion fires; call clearIncorrectValues() after sync instead of custom prune-only logic. Remove model.onValueChanged once covered.

File: lib/survey-features/advanced-carry-forward/infrastructure/registry.ts [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+
+    originalDependsOn: property.dependsOn,
+
+  };
+
+  patchedNativeProperties.set(property, snapshot);
+
+
+
+  property.visibleIf = (obj) => {
```
Comment (PeterMarinov) [inline lib/survey-features/advanced-carry-forward/infrastructure/registry.ts]: Native choices / choicesFromQuestion / choicesByUrl properties are patched globally; reset exists only for tests. Risk of stale wrappers under HMR or double-init.

Fix: Document one-time init contract; add dev assert when re-init detected; keep test reset. Ideally, we need to use this approach - https://surveyjs.io/form-library/documentation/customize-question-types/add-custom-properties-to-a-form#override-default-property-values and hide via the visibleIf override of the native property

File: lib/survey-features/data-lists/infrastructure/registry.ts [hunk 17]
```diff
@@ -21,9 +21,12 @@ export function registerDataListGlobals(): void {
     displayName: "Data list",
     category: "choices",
     visibleIndex: 0,
-    dependsOn: "choicesFromQuestion",
-    visibleIf: (obj: { choicesFromQuestion?: unknown }) => {
-      return !obj.choicesFromQuestion;
+    dependsOn: ["choicesFromQuestion", "advancedCarryForwardEnabled"],
```
Comment (PeterMarinov) [inline lib/survey-features/data-lists/infrastructure/registry.ts]: Runtime enablement, Creator visibility, and data-list hiding can drift. Data lists uses magic string 'advancedCarryForwardEnabled' instead of a shared constant.

Fix: Extract lib/survey-features/infrastructure/choice-source-mutual-exclusion.ts with one predicate; import everywhere.

File: lib/survey-features/advanced-carry-forward/use-cases/sync-carry-forward-target.ts [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+    for (const choice of source.choices ?? []) {
+      if (!index.has(choice.value)) {
+        index.set(choice.value, choice);
+      }
+    }
+  }
```
Comment (PeterMarinov) [inline lib/survey-features/advanced-carry-forward/use-cases/sync-carry-forward-target.ts]: this compared to vs sync-carry-forward-target.ts:62-72 & loop-utils.ts:218 offers diverged parsing of values. Dedup/enrichment uses raw choice.value; priority split and selection pruning use String(value). Mixed-type values across sources can mis-order priority items or fail to prune selections.

Fix: Use shared normalizeChoiceKey in [lib/utils/survey/choice-values.ts](hub/lib/utils/survey/choice-values.ts) backed by [parseScalarString](hub/lib/utils/type-parsers.ts) everywhere dedup, priority, prune, and enrichment touch choice values.

File: lib/survey-features/advanced-carry-forward/constants.ts [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+export const ADVANCED_CARRY_FORWARD_ENABLED_PROPERTY =
+  'advancedCarryForwardEnabled';
+export const ADVANCED_CARRY_FORWARD_SOURCES_PROPERTY =
+  'advancedCarryForwardSources';
+export const ADVANCED_CARRY_FORWARD_MODE_PROPERTY = 'advancedCarryForwardMode';
+export const ADVANCED_CARRY_FORWARD_PRIORITY_ITEMS_PROPERTY =
+  'advancedCarryForwardPriorityItems';
+export const ADVANCED_CARRY_FORWARD_MAX_CHOICES_PROPERTY =
+  'advancedCarryForwardMaxChoices';
+
```
Comment (PeterMarinov) [inline lib/survey-features/advanced-carry-forward/constants.ts]: Since we store this in the survey JSON, I recommend 2 prone approach:
1. We offer a strategy of keeping jsonKeys short to allocate less space
2. We add `edx` prefix to avoid naming collision with future sjs property names

`advancedCarryForwardPriorityItems` => `edxCarryForwardPriorityItems`

Maybe second point is more important as it also allows for instant identification of Endatix customizations. Let me know what you think of this.

Has been applied to dataLists

**Review by PeterMarinov — DISMISSED** [review]
This is looking great. The functionality offers robust customization engine with test coverage and can be merged as it is.

Still, I think there must be one iteration focused on resolving future problems and it's better to be done now.

4 key reason why not merge:
1. We try to use more of the native `question_baseselect.ts`'s features like tracking dependent questions instead of doing custom tracking state machine. This way we solve for future performance and step on already tested and efficient "choicesFromQuestions" mechanics instead of hooking on every prop change via `onPropertyChanged`
2. We update our docs for even cleaner customization path
3. Renaming the properties - better to be done now to avoid breaking changes
4. Suggesting we shorten the vertical slice name to "carry-forward" (from "advanced-carry-forward" and avoid rename after this is in production

I will try addressing the comments `chore/h706-refactor-code` to show how this might look and let you decide

File: lib/survey-features/carry-forward/utils/carry-forward-graph.ts
Comment (PeterMarinov) [inline lib/survey-features/carry-forward/utils/carry-forward-graph.ts:1]: We put functional (verb) bases use cases that encapsulate core feature logic. Utils we put outside e.g. `utils.ts`. Best to move it out

**Review by PeterMarinov — APPROVED** [review]
Approved. All issues and comments were addressed with merging #716 & #714
